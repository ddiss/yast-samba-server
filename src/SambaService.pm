# File:		modules/SambaService.ycp
# Package:	Configuration of samba-server
# Summary:	Data for configuration of samba-server, input and output functions.
# Authors:	Stanislav Visnovsky <visnov@suse.cz>
#		Martin Lazar <mlazar@suse.cz>
#
# $Id$
#
# Representation of the configuration of samba-server.
# Input and output routines.


package SambaService;

use strict;
use Switch 'Perl6';
use Data::Dumper;

use YaST::YCP qw(:DATA :LOGGING);
use YaPI;

textdomain "samba-server";
our %TYPEINFO;

BEGIN {
YaST::YCP::Import("Service");
}

# Data was modified?
my $Modified = 0;

# Is smb and nmb service enabled? 
my $Service = 0;

# Data was modified?
BEGIN{ $TYPEINFO{GetModified} = ["function", "boolean"] }
sub GetModified {
    my ($self) = @_;
    return $Modified;
};

# Export
BEGIN{$TYPEINFO{Export}=["function", "any"]}
sub Export {
    my ($self) = @_;
    return $Service ? "Enabled" : "Disabled";
}

# Import
BEGIN{$TYPEINFO{Import}=["function", "void", "any"]}
sub Import {
    my ($self, $any) = @_;
    $any = "No" unless $any;
    $self->SetServiceAutoStart(($any =~ /^(1|Enabled?|True|Yes)$/i) ? 1 : 0);
}

# Read
BEGIN{$TYPEINFO{Read}=["function", "boolean"]}
sub Read {
    my ($self) = @_;
    $Service = Service->Enabled("smb");
    $Modified = 0;
    return 1;
}

# Write
# return true on succes
BEGIN{$TYPEINFO{Write}=["function", "boolean"]}
sub Write {
    my ($self) = @_;
    y2debug("Samba service if ". ($Service ? "enabled" : "disabled"));
    Service->Adjust("nmb", $Service ? "enable" : "disable") or return 0;
    Service->Adjust("smb", $Service ? "enable" : "disable") or return 0;
    $Modified = 0;
    return 1
}

# Adjust SAMBA server services (smb and nmb).
BEGIN{$TYPEINFO{SetServiceAutoStart} = ["function", "void", "boolean"]}
sub SetServiceAutoStart {
    my ($self, $on) = @_;
    unless (($on && $Service) || (!$on && !$Service)) {
	$Service = $on ? 1 : 0;
	$Modified = 1;
    }
}

# Get SAMBA server services (smb and nmb) status.
BEGIN{$TYPEINFO{GetServiceAutoStart} = ["function", "boolean"]}
sub GetServiceAutoStart {
    return $Service ? 1 : 0;
}

# Start/Stop SAMBA server daemons (smb and nmb) NOW.
# return @integer	0 on succes, -1 if cannot start, -2 if cannot reload, -3 if cannot stop
BEGIN{$TYPEINFO{StartStopNow}=["function", "boolean", "boolean"]};
sub StartStopNow {
    my ($self, $on) = @_;
    
    foreach("nmb", "smb") {
	if ($on) {
    	    # check, if the services run
	    if (Service->Status($_)) {
		# the service does not run => start it
		unless (Service->Start($_)) {
		    y2error("Service::Start($_) failed");
		    return 0;
		}
	    } else {
		# the service runs => relaod it
		unless (Service->RunInitScript($_, "restart")) {
		    y2error("Service::RunInitScript($_, 'restart') failed");
		    return 0;
		}
	    }
	} else {
	    # turn services off
	    unless (Service->Status($_)) {
		unless (Service->Stop($_)) {
		    y2error("Service::Stop($_) failed");
		    return 0;
		}
	    }
	}
    }
    
    return 1;
}

8;
