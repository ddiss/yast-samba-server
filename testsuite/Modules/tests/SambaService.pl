#!/usr/bin/perl

use SambaService;
use Data::Dumper;

## test agent
=x
use YaST::YCP qw(:DATA);
YaST::YCP::Import("Testsuite");
my $r_ok = {target=>{tmpdir=>"/tmp"}};
my $w_ok = {target=>{string=>Boolean(1)}};
my $w_err = {target=>{string=>Boolean(0)}};
my $e_err = {target=>{bash=>8, remove=>8, bash_output=>{exit=>8}}};
my $e_ok = {target=>{bash=>0, remove=>0, bash_output=>{exit=>0}}};
=cut
=disable
## fake modules
sub Service::Enabled {$enabled}
sub Service::Status {$status}
sub Service::Start {print "Service->Start($_[1])\n"; $start}
sub Service::Stop {print "Service->Stop($_[1])\n"; $stop}
sub Service::RunInitScript {print "Service->RunInitScript($_[1], $_[2])\n"; $restart}
sub Service::Adjust {print "Service->Adjust($_[1], $_[2])\n"; $adjust}


## Export/Import()
SambaService->Import();
print Dumper(
    SambaService->Export(),
    !SambaService->GetModified());

SambaService->Import("Enabled");
print Dumper(
    SambaService->Export(),
    !SambaService->GetModified());

SambaService->Import("Off");
print Dumper(
    SambaService->Export(),
    !SambaService->GetModified());


## Set/GetServiceAutostart(); GetModified()
$SambaService::Modified=0;
$SambaService::Service=0;
SambaService->SetServiceAutoStart(); # undef => ON
print Dumper(
    SambaService->GetServiceAutoStart(),
    SambaService->GetModified());

$SambaService::Modified=0;
$SambaService::Service=0;
SambaService->SetServiceAutoStart(0);
print Dumper(
    !SambaService->GetServiceAutoStart(),
    !SambaService->GetModified());

$SambaService::Modified=0;
$SambaService::Service=1;
SambaService->SetServiceAutoStart(0);
print Dumper(
    !SambaService->GetServiceAutoStart(),
    SambaService->GetModified());

$SambaService::Modified=0;
$SambaService::Service=1;
SambaService->SetServiceAutoStart(1);
print Dumper(
    SambaService->GetServiceAutoStart(),
    !SambaService->GetModified());


## Read()/Write()
$SambaService::Modified=1;
$enabled = 0;
print Dumper(
    SambaService->Read(),
    !SambaService->GetModified(),
    !SambaService->GetServiceAutoStart());
    
$SambaService::Modified=1;
$SambaService::Service=1;
$adjust = 1;
print Dumper(
    SambaService->Write(),
    !SambaService->GetModified());

$SambaService::Modified=0;
print Dumper(SambaService->Write());

$SambaService::Modified=1;
$SambaService::Service=0;
$adjust = 0;
print Dumper(!SambaService->Write());


## StartStopNow()
$status = 1;
$start = 0;
print Dumper(!SambaService->StartStopNow(1));

$status = 1;
$start = 1;
print Dumper(SambaService->StartStopNow(1));

$status = 0;
$restart = 0;
print Dumper(!SambaService->StartStopNow(1));

$status = 0;
$restart = 1;
print Dumper(SambaService->StartStopNow(1));

$status = 0;
$stop = 0;
print Dumper(!SambaService->StartStopNow(0));

$status = 0;
$stop = 1;
print Dumper(SambaService->StartStopNow(0));

$status = 1;
print Dumper(SambaService->StartStopNow(0));
