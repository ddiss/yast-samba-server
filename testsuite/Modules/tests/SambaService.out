$VAR1 = 'Disabled';
$VAR2 = 1;
$VAR1 = 'Enabled';
$VAR2 = 1;
$VAR1 = 'Disabled';
$VAR2 = 1;
$VAR1 = 1;
$VAR2 = 1;
$VAR1 = 1;
$VAR2 = 1;
$VAR1 = 1;
$VAR2 = 1;
$VAR1 = 1;
$VAR2 = 1;
$VAR1 = 1;
$VAR2 = 1;
$VAR3 = 1;
Service->Adjust(nmb, enable)
Service->Adjust(smb, enable)
$VAR1 = 1;
$VAR2 = 1;
$VAR1 = 1;
Service->Adjust(nmb, disable)
Service->Adjust(smb, disable)
$VAR1 = 1;
Service->Start(nmb)
Service->Start(smb)
$VAR1 = 1;
Service->Start(nmb)
Service->Start(smb)
$VAR1 = 1;
Service->RunInitScript(nmb, restart)
Service->RunInitScript(smb, restart)
$VAR1 = 1;
Service->RunInitScript(nmb, restart)
Service->RunInitScript(smb, restart)
$VAR1 = 1;
Service->Stop(nmb)
Service->Stop(smb)
$VAR1 = 1;
Service->Stop(nmb)
Service->Stop(smb)
$VAR1 = 1;
$VAR1 = 1;
